src/vulkan-renderer/wrapper/cpu_texture.cpp

```diff
@@ -0,0 +1,87 @@
+﻿#include "inexor/vulkan-renderer/wrapper/cpu_texture.hpp"
+
+#define STB_IMAGE_IMPLEMENTATION
+#include <spdlog/spdlog.h>
+#include <stb_image.h>
+
+#include <array>
+#include <vector>
+
+namespace inexor::vulkan_renderer::wrapper {
+
+CpuTexture::CpuTexture() : m_file_name("default texture"), m_name("default texture") {
+    generate_error_texture_data();
+}
+
+CpuTexture::CpuTexture(const std::string &file_name, const std::string &name) : m_file_name(file_name), m_name(name) {
+    assert(!m_file_name.empty());
+    assert(!m_name.empty());
+
+    spdlog::debug("Loading texture file {}.", file_name);
+
+    // Load the texture file using stb_image library.
+    // Force stb_image to load an alpha channel as well.
+    stbi_uc *texture_data = stbi_load(file_name.c_str(), &m_texture_width, &m_texture_height, nullptr, STBI_rgb_alpha);
+
+    if (texture_data == nullptr) {
+        spdlog::error("Could not load texture file {} using stbi_load!  Falling back to error texture.", file_name);
+        generate_error_texture_data();
+    } else {
+
+        // TODO: We are currently hard coding the number of channels with STBI_rgb_alpha.
+        //       Eventually, we probably need to pass this information into this class from
+        //       a higher level class - like a material loader class.
+        //       So, as an example, if the material loader is loading a normal map, we know
+        //       we need to tell this class to load a 3 channel texture.  If it can, great.
+        //       If it can not, then this class probably needs to load a 3 channel error texture.
+        m_texture_channels = 4;
+
+        // TODO: We are currently only supporting 1 mip level.
+        m_mip_levels = 1;
+
+        const int texture_data_size{m_texture_width * m_texture_height * m_texture_channels};
+        m_texture_data.resize(texture_data_size);
+        std::copy(texture_data, texture_data + texture_data_size, m_texture_data.begin());
+
+        spdlog::debug("Texture dimensions: width: {}, height: {}, channels: {} mip levels: {}.", m_texture_width,
+                      m_texture_height, m_texture_channels, m_mip_levels);
+    }
+
+    // Discard the texture data.
+    stbi_image_free(texture_data);
+}
+
+CpuTexture::CpuTexture(CpuTexture &&other) noexcept
+    : m_file_name(std::move(other.m_file_name)), m_name(std::move(other.m_name)),
+      m_texture_width(other.m_texture_width), m_texture_height(other.m_texture_height),
+      m_texture_channels(other.m_texture_channels), m_mip_levels(other.m_mip_levels),
+      m_texture_data(std::move(other.m_texture_data)) {}
+
+CpuTexture::~CpuTexture() {}
+
+void CpuTexture::generate_error_texture_data() {
```

Comment:
Do you think it would be possible to make this a static function that returned a new `CpuTexture` object? I feel like it would be much nicer than having a member function that changed state after construction.

Reply:
Thought about that too, but if a texture is not loaded the error texture is automatically applied. also the function is private.

Reply:
I am not too concerned about this, @authenticate feel free to mark this as resolved.